prime-router/src/main/kotlin/azure/Event.kt

```diff
@@ -75,23 +82,90 @@ abstract class Event(val eventAction: EventAction, val at: OffsetDateTime?) {
+            if (parts.size < 3 || parts.size > 7) error("Internal Error: bad event format")
             val action = EventAction.parseQueueMessage(parts[1])
-            val after = parts.getOrNull(3)?.let { OffsetDateTime.parse(it) }
             return when (parts[0]) {
                 ReportEvent.eventType -> {
+                    val after = parts.getOrNull(3)?.let { OffsetDateTime.parse(it) }
                     val reportId = UUID.fromString(parts[2])
                     ReportEvent(action, reportId, after)
                 }
                 ReceiverEvent.eventType -> {
+                    val after = parts.getOrNull(3)?.let { OffsetDateTime.parse(it) }
                     ReceiverEvent(action, parts[2], after)
                 }
+                ProcessEvent.eventType -> {
+                    val after = parts.getOrNull(6)?.let { OffsetDateTime.parse(it) }
+                    val reportId = UUID.fromString(parts[2])
+                    val options = Options.valueOf(parts[3])
+
+                    // convert incoming serialized routeTo string into List<String>
+                    val routeTo = parts.getOrNull(5)?.let {
+                        it.split(ROUTE_TO_SEPARATOR)
+                    } ?: emptyList()
+
+                    // convert incoming defaults serialized string to Map<String,String>
+                    val defaults = parts.getOrNull(4)?.let {
+                        it.split(',').associate { pair ->
+                            val defaultParts = pair.split(DEFAULT_SEPARATOR)
+                            Pair(defaultParts[0], defaultParts[1])
+                        }
+                    } ?: emptyMap()
+
+                    ProcessEvent(action, reportId, options, defaults, routeTo, after)
+                }
                 else -> error("Internal Error: invalid event type: $event")
             }
         }
     }
 }
 
+/**
+ * An event that goes on and is read off of the 'process' queue.
+ * @param reportId The report ID to be processed
+ * @param options Options passed in on the initial call
+ * @param defaults Defaults passed in on the initial call
+ * @param routeTo Route to overrides passed in on the initial call
+ * @param at Allows future scheduling of the queue message
+ * @param retryToken For retrying a process queue message
+ */
+class ProcessEvent(
+    eventAction: EventAction,
+    val reportId: UUID,
+    val options: Options,
+    val defaults: Map<String, String>,
+    val routeTo: List<String>,
+    at: OffsetDateTime? = null,
+    val retryToken: RetryToken? = null
+) : Event(eventAction, at) {
+    override fun toQueueMessage(): String {
+        // turn the defaults and route to into strings that can go on the process queue
+        val routeToQueueParam = routeTo.let { it.joinToString(",") }
+        val defaultsQueueParam = defaults.let { it.map { pair -> "${pair.key}:${pair.value}" }.joinToString(",") }
```

Comment:
Ditto here about the let. Because defaults is not null, you don't need to wrap in a let.

Reply:
Updated